tools/topology/topology1/kernel_dependent/v5.18/CMakeLists.txt

```
@@ -0,0 +1,49 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+# Those topologies are built from topology in the parent directory
+set(TPLGS_UP
+	## HDaudio codec topologies
+	"sof-hda-generic\;sof-hda-generic\;-DDEEP_BUFFER\;-DCHANNELS=0\;-DHSPROC=volume\;-DDYNAMIC=1"
+	"sof-hda-generic\;sof-hda-generic-1ch\;-DDEEP_BUFFER\;-DCHANNELS=2\;-DHSPROC=volume\;-DDMICPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_48khz.m4\;-DDMIC16KPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_16khz.m4\;-DDYNAMIC=1"
+	"sof-hda-generic\;sof-hda-generic-2ch\;-DDEEP_BUFFER\;-DCHANNELS=2\;-DHSPROC=volume\;-DDMICPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_48khz.m4\;-DDMIC16KPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_16khz.m4\;-DDYNAMIC=1"
+	"sof-hda-generic\;sof-hda-generic-3ch\;-DDEEP_BUFFER\;-DCHANNELS=4\;-DHSPROC=volume\;-DDMICPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_48khz.m4\;-DDMIC16KPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_16khz.m4\;-DDYNAMIC=1"
+	"sof-hda-generic\;sof-hda-generic-4ch\;-DDEEP_BUFFER\;-DCHANNELS=4\;-DHSPROC=volume\;-DDMICPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_48khz.m4\;-DDMIC16KPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_16khz.m4\;-DDYNAMIC=1"
+	## end HDaudio codec topologies
+)
+
+add_custom_target(kernel_dependent_v5_18_topologies1 ALL)
```

Shouldn't this be?
```suggestion
add_custom_target(kernel_dependent_v5_18_topologies1)
add_dependencies(topologies1 kernel_dependent_v5_18_topologies1)
```
In other words, shouldn't these be built every time other topologies1 are built?

---

this was taken from the development directory, I have no idea what CMake does and these topologies are built at the same time as the others, I don't know what the issue is.

---

CMake/Make/ninja/etc. all have a tree of dependencies, they all work the same at a high-level. You never have to build ALL, you can always ask for any specific node in tree. Asking for a specific node (e.g. "topologies1") builds only that node and all its leaves / dependencies and nothing else. The question here is under what node should this new leaf be.

This also defines what can be built in parallel (which makes a huge speed difference here) and if not then in what order. The order matters even more here because topologiesv2 overrides v1, see below.

---

I don't understand any of this. The same code with ALL is used everywhere already:

````
git grep add_custom_target | grep ALL
tools/topology/CMakeLists.txt:add_custom_target(topologies ALL)
tools/topology/topology1/CMakeLists.txt:add_custom_target(topologies1 ALL
tools/topology/topology1/development/CMakeLists.txt:add_custom_target(dev_topologies1 ALL)
tools/topology/topology1/dsp_enhancement/CMakeLists.txt:add_custom_target(dsp_topologies1 ALL)
tools/topology/topology1/kernel_dependent/v5.18/CMakeLists.txt:add_custom_target(kernel_dependent_v5_18_topologies1 ALL)
````

---

@marc-hb ping on this, I have no idea what change you are asking for?

---

> I don't understand any of this. 

Do you understand "tree of dependencies"? For instance, have you noticed that `.c` files can be compiled to `.o` files in parallel without waiting for each other but that creating `vmlinuz` has to wait for `.o` files to be created? Unless  `.o` files which are part of a module, in which case asking for `vmlinuz` won't build them, you have to ask for `modules` instead?

> The same code with ALL is used everywhere already

You found only the top-level targets built by default when you don't say what you want to build. `topologies2` is not in that list. `topologies2` not a top-level target, it's  a dependency of other targets. Whether `topologies2` is built or not depends on what is requested.

Other example: if you ask to build the top-level target `topologies` with `./scripts/build-tools.sh -T` then `dev_topologies` are NOT built at all. `./scripts/build-tools.sh -h` warns about that.



The initial question is just: when should these topologies be built? Should they be built even when the user does not explicitly request them? As part of some other target / "group" of topologies?

---

Can we please stop talking about topologies2? This PR is about topology1, I have no idea what we're having this debate on topology2. Different problem and different PR.

My point is that ONLY on topology1 and I don't see why you are asking for a change when all other existing topology1. Again see the list that I provided:

````
git grep add_custom_target | grep ALL
tools/topology/CMakeLists.txt:add_custom_target(topologies ALL)
tools/topology/topology1/CMakeLists.txt:add_custom_target(topologies1 ALL
tools/topology/topology1/development/CMakeLists.txt:add_custom_target(dev_topologies1 ALL)
tools/topology/topology1/dsp_enhancement/CMakeLists.txt:add_custom_target(dsp_topologies1 ALL)
tools/topology/topology1/kernel_dependent/v5.18/CMakeLists.txt:add_custom_target(kernel_dependent_v5_18_topologies1 ALL)
````

---

topology2 was just one example, I gave another. You're adding a new dependency in the tree while trying hard not to look at the existing dependencies, so I'm afraid we're stuck. Dependencies is how any build system work, we can't opt out. What gets built and released, how and when is not just an implementation detail, it's part of the "user interface".

---

@marc-hb can the topology build be fixed/improved after this PR is merged ? Is this something you can align with your vision of how it should work ?
Fwiw, testing the deep buffer is high priority, so it's important we get this in for dog fooding.

---

As one should always do when text communication get stuck, I picked up the phone and had productive chat with @plbossart 

> Fwiw, testing the deep buffer is high priority, so it's important we get this in for dog fooding.

Without my dependency suggestion above, deep buffer topologies may or may not be built. More specifically, deep buffer topologies will never be built if any option is passed to `./scripts/build-tools.sh`.

With my dependency suggestion above, deep buffer topologies will always be built with `./scripts/build-tools -T` (and also built when no option is passed). I confirmed that this is what @plbossart wants to happen.

I tested all combinations and verified everything I just wrote.

tl;dr: my suggestion should be applied.

---

@marc-hb if good can you approve. Do you PRs need to be applied before this PR ?